### ISSUE TITLE
iter_cloned_collect false positive with custom From/IntoIterator impl

### ISSUE THREAD

**Issue by Dentosal** [rust-lang/rust-clippy#9119]
### Summary

When `FromIterator` is implemented for a type inside an another iterator, `.collect()`  is able to do run the inner iterator as well. Clippy doesn't recognize this case, and instead assumes that `.iter().cloned().collect()` to a `Vec<_>` always has the same resulting type than `to_vec()`.  This also occurs with `.iter().copied().collect()`.

### Lint Name

iter_cloned_collect and iter_cloned_collect

### Reproducer

I tried this code:

```rust
use std::iter;

#[derive(Clone)]
struct Example(u16);

impl iter::FromIterator<Example> for Vec<u8> {
    fn from_iter<T>(iter: T) -> Self
    where
        T: IntoIterator<Item = Example>,
    {
        iter.into_iter().flat_map(|e| e.0.to_le_bytes()).collect()
    }
}

fn main() {
    let examples = vec![Example(1), Example(0x1234)];
    let encoded: Vec<u8> = examples.iter().cloned().collect();
    assert_eq!(encoded, vec![0x01, 0x00, 0x34, 0x12]);
}

```

I saw this happen:

```
warning: called `iter().copied().collect()` on a slice to create a `Vec`. Calling `to_vec()` is both faster and more readable
  --> src/main.rs:17:36
   |
17 |     let encoded: Vec<u8> = examples.iter().copied().collect();
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `.to_vec()`
   |
   = note: `#[warn(clippy::iter_cloned_collect)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_cloned_collect
```

I expected to see this happen:

No warning, as the `to_vec` cannot do the required conversion.

### Version

```text
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.62.0
LLVM version: 14.0.5
```

### Additional Labels

@rustbot label +l-suggestion-causes-error
